Proposal RFC: Java Agents and More Layers

coollog opened the pull request:
**Update:** This proposal will probably be retired in favor of the incubating `src/main/jib` convention. (#489)

Request for comments. Approval upon merge.

Preview at https://github.com/GoogleContainerTools/jib/blob/more-layers-proposal/proposals/java_agents_and_more_layers.md

@linux-china
@saturnism

Fixes #378 
Fixes #213 (closes in favor of this)
Regards #403

briandealwis commented on proposals/java_agents_and_more_layers.md:
```diff
@@ -0,0 +1,280 @@
+
+### Alternative (Rejected) Proposal
+
+The alternative proposal was rejected because we deemed that layering should be an implementation detail that should not be exposed to the user.
```
Maybe call this out with `**The alternative proposal was rejected …**`?

briandealwis commented on proposals/java_agents_and_more_layers.md:
```diff
@@ -0,0 +1,280 @@
+    <from>another/file</from>
+    <to>/another/path/on/image</to>
+  </addFile>
+</addFiles>
```
Oh jeepers 🤦‍♂️  There does [seem to be a way to support relatively arbitrary XML](https://stackoverflow.com/a/32800814/600339) but it's not pretty and I don't think it's warranted by this situation.

loosebazooka replied:
yeah that might be a little un-maven-y

briandealwis commented on proposals/java_agents_and_more_layers.md:
````diff
@@ -58,7 +58,7 @@ addFile 'path/to/file', '/path/on/image'
 ```
 
-All of the files would be added to a single new layer.
+*Directories will be handled using [rsync](https://linux.die.net/man/1/rsync) source-path semantics, where, for example, `addFile 'path/to/directory', '/path/on/image'` would put all the files in `path/to/directory` into the directory `/path/on/image/directory`, but `addFile 'path/to/directory', '/path/on/image'` (trailing slash) would put all the files into the directory `/path/on/image`.* 
````
Second example is missing the trailing slash!

coollog replied:
Oops, will add that.

coollog commented on proposals/java_agents_and_more_layers.md:
```diff
@@ -0,0 +1,280 @@
+
+The configuration for adding additional files (including Java agents) would look something like:
+
+*(The exact configuration is to be decided before this PR is merged. Alternative naming for this parameter include: `additionalFiles`, `extraFiles`, and `copyFiles`.)*
```
Actually, I was just thinking and maybe we should have it be the semantics of [Dockerfile `COPY`](https://docs.docker.com/engine/reference/builder/#copy), where essentially:

- if `from` is a directory, only the contents are copied, and not the directory itself
- if `to` ends with a trailing slash, it is considered a directory. Otherwise, it is considered a regular file.
- if `from` refers to multiple files (like the contents of a directory), `to` must end with a trailing slash, and the `from` files will be added to that directory.

We won't support glob matching though, at least not until someone says they need it.

briandealwis replied:
Using Docker's COPY semantics makes sense.

loosebazooka commented:
I just tested `copies`, and it does become `copy` (maven knows how to do this)
```
<copies>
  <copy>asdf</copy>
</copies>
```
but I'm not sure if this actually reads well.

TadCordle commented:
Are there any plans for this proposal, or should we just close this PR?

coollog commented:
I think this will be good to have as a record in the proposals (and as a reference for the design behind the current solution). I'll touch this up with our current solution (`src/main/jib`) and move the previous solutions to sections about rejected proposals.

coollog commented:
Changed the main proposal to regard `src/main/jib`, moving the previous one to rejected. Also moved this to `archives` since `src/main/jib` is already implemented.

briandealwis commented on proposals/archives/java_agents_and_more_layers.md:
```diff
@@ -0,0 +1,292 @@
+
+The directory should also be able to be configured to override the `src/main/jib` default.
+
+### Alternative (Rejected) Proposal
```
Would be good to have a `## Alternative Proposals` here as they otherwise fall under the `## Proposal`.
